### ISSUE TITLE
Command line length limit for gcc linking

### ISSUE THREAD

**Issue by kayhayen** [SCons/scons#3348]
* Version of SCons

3.0.4

* Version of Python
2.7 through 3.7

* Which python distribution if applicable (python.org, cygwin, anaconda, macports, brew,etc)
All

* How you installed SCons
apt-get install on Debian, and inline copy of source code.

* What Platform are you on? (Linux/Windows and which version)
Windows 10, Debian Linux and others. Nuitka is used by various people.

* How to reproduce your issue?  Please include a small self contained reproducer. Likely a SConstruct should do for most issues.
I lack a SCCE, and did the fix for Nuitka without one, but compiling Tensorflow on Linux with Nuitka seems to do it. It is however not necessary to see that the linker command is unbound, unlike with mslink.

* How you invoke scons (The command line you're using "scons --flags some_arguments")

We do only these:

```python
        "--warn=no-deprecated",
        # Don't load "site_scons" at all.
        "--no-site-dir",
```

To work around the issue, I have added code in Nuitka like this:

```python
if gcc_mode:
    tmp_linker_filename = os.path.join(source_dir, "@sources.tmp")

    env["LINKCOM"] = env["LINKCOM"].replace(
        "$SOURCES", "@%s" % env.get("ESCAPE", lambda x: x)(tmp_linker_filename)
    )

    with open(tmp_linker_filename, "w") as tmpfile:
        for filename in source_files:
            filename = ".".join(filename.split(".")[:-1]) + ".o"
            tmpfile.write('"%s"\n' % filename)

        tmpfile.write(env.subst("$SOURCES"))

Default(target)
```

I didn't have the skills to produce a proper scons patch here, as they way it's done for mslink seems to involve more Scons knowledge than I have. But I think it's still valuable to point out that the flaw exists in Scons.

The issue occurs most often with MingGW64 due to the lower limits on Windows I think, but people have managed to hit it with gcc on Linux too, where 32k is the limit I believe.

**Comment by bdbaddog** [comment]
Which platform are you having this issue on? (Not which platforms do the software build on)

There's already logic in SCons to deal with lines to long by placing in a temp file.

Also I guess you ignored the request to bring this issue to the scons users mailing list.

Try:
    env["LINKCOM"] = "${TEMPFILE %s"%env['LINKCOM']
    env['TEMPFILEPREFIX'] = '@'

**Comment by acmorrow** [comment]
I think though that by default, the long line stuff is only applied for Windows. At least for MongoDB, we had to do some work to apply it more generally:

- https://github.com/mongodb/mongo/blob/master/SConstruct#L769-L775
- https://github.com/mongodb/mongo/blob/master/SConstruct#L1482-L1500

**Comment by bdbaddog** [comment]
> I think though that by default, the long line stuff is only applied for Windows. At least for MongoDB, we had to do some work to apply it more generally:
> 
>     * https://github.com/mongodb/mongo/blob/master/SConstruct#L769-L775
> 
>     * https://github.com/mongodb/mongo/blob/master/SConstruct#L1482-L1500

Yes. And the logic above should effectively do what you've done in your build for LINKCOM which was the only one mentioned

**Comment by kayhayen** [comment]
I am having the issue with gcc on all platforms, @bdbaddog and it depends on the linker used. Only with MSVC I am seeing the temp file trick being done (and for we have another fix in another issue) and it seems to be pretty involved, and non generic.

For my naive eyes, putting things in a batch file ought to not help. Instead the `@file` must be used, similar to what is done in mslink.

**Comment by bdbaddog** [comment]
> I am having the issue with gcc on all platforms, @bdbaddog and it depends on the linker used. Only with MSVC I am seeing the temp file trick being done (and for we have another fix in another issue) and it seems to be pretty involved, and non generic.
> 
> For my naive eyes, putting things in a batch file ought to not help. Instead the `@file` must be used, similar to what is done in mslink.

Did you try the code I posted to see if it works for your issue?

**Comment by kayhayen** [comment]
@bdbaddog When I did it, it was giving me errors:
```console
scons: *** [Mini.bin] SyntaxError `invalid syntax (<string>, line 1)' trying to evaluate `${TEMPFILE $LINK -o $TARGET $LINKFLAGS $__RPATH $SOURCES $_LIBDIRFLAGS $_LIBFLAGS}'
```
The code I used, fixing what I think it s typo:

```python
    env["LINKCOM"] = "${TEMPFILE %s}"%env['LINKCOM']
    env['TEMPFILEPREFIX'] = '@'
```
